_posts/2019-06-26-kafka-streams-applications-with-quarkus-and-microprofile.adoc

```diff
@@ -0,0 +1,361 @@
+        // set up Kafka Streams config, e.g. sourced from application.properties
+        Properties props = new Properties();
+        // props.put(..., ...);
+
+        // set up the stream topology
+        StreamsBuilder builder = new StreamsBuilder();
+        // builder.stream(...)
+
+        streams = new KafkaStreams(builder.build(), props);
+        streams.start();
+    }
+
+    void onStop(@Observes ShutdownEvent event) {
+        streams.close();
+    }
+}
+----
+
+For the very common requirement that you'd like to serialize and deserialize Java types used in the streaming pipeline into/from JSON
+(e.g. when materializing them in a state store),
+the Quarkus Kafka Streams extension provides the class `io.quarkus.kafka.client.serialization.JsonbSerde`.
+It's a `Serde` implementation based on JSON-B:
+
+[source, java]
+----
+...
+JsonbSerde<WeatherStation> weatherStationSerde = new JsonbSerde<>(WeatherStation.class);
+
+GlobalKTable<Integer, WeatherStation> stations = builder.globalTable(
+    "weather-stations",
+    Consumed.with(Serdes.Integer(), weatherStationSerde)
+);
+...
+----
+
+== Running Native
+
+Based on Kafka's notion of topic partitioning, Kafka Streams applications can easily be scaled out:
+the load will be shared amongst multiple instances of the application,
+each processing just a subset of the partitions of the input topic(s).
+
+Running Quarkus applications in native code via GraalVM comes in very handy for that:
+besides the very fast start-up time, the applications will use significantly less memory when compiled to native code.
+This lets you spin up many instances of a Quarkus-based Kafka Streams pipeline in parallel,
+in a very memory-efficient way.
+
+The extension takes care of everything needed for building native Kafka Streams applications,
+for instance it makes sure that the right RocksDB native libraries are added to the application image.
+All you need to do is to specify the `enableJni` option for the Quarkus Maven plug-in,
+allowing those native libraries to be invoked via JNI:
+
+[source, xml]
+----
+<plugin>
+    <groupId>io.quarkus</groupId>
+    <artifactId>quarkus-maven-plugin</artifactId>
+    <executions>
+        <execution>
+            <goals>
+                <goal>native-image</goal>
+            </goals>
+            <configuration>
+                <enableJni>true</enableJni>
+            </configuration>
+        </execution>
+    </executions>
+</plugin>
+----
+
+When using the `JsonbSerde` for unmarshalling JSON into corresponding Java objects,
+those types must be marked with the `@RegisterForReflection` annotation,
+so that they can to be instantiated reflectively by JSON-B in native mode:
+
+[source, java]
+----
+@RegisterForReflection
+public class WeatherStation {
+
+    public int id;
+    public String name;
+}
+----
+
+Then build the application using the `native` profile
+(note this requires GraalVM to be installed on your system):
```

Comment:
I would add a link to https://quarkus.io/guides/building-native-image-guide .

Reply:
Good idea.